// sanity-plugin-asset-source-unsplash/lib/index.js
import { useClient, definePlugin } from "sanity";
import { jsxs, jsx } from "react/jsx-runtime";
import { SearchIcon } from "@sanity/icons";
import { Card, Text, TextInput, Dialog, Stack, Flex, Spinner } from "@sanity/ui";
import React, { useCallback, useEffect } from "react";
import InfiniteScroll from "react-infinite-scroll-component";
import PhotoAlbum from "react-photo-album";
import { concat, defer, BehaviorSubject } from "rxjs";
import { withLatestFrom, debounceTime, distinctUntilChanged, switchMap, map } from "rxjs/operators";
import { styled } from "styled-components";
function UnsplashIcon() {
  return /* @__PURE__ */ jsxs("svg", { role: "img", viewBox: "0 0 25 25", width: "1em", height: "1em", fill: "currentColor", children: [
    /* @__PURE__ */ jsx("title", {}),
    /* @__PURE__ */ jsx("path", { d: "M9 9V4h7v5H9Zm7 3h5v9H4v-9h5v5h7v-5Z" })
  ] });
}
const fetchSearch = (client, query, page, perPage) => defer(
  () => client.observable.request({
    url: `/addons/unsplash/search/photos?query=${encodeURIComponent(
      query
    )}&page=${page}&per_page=${perPage}`,
    withCredentials: !0,
    method: "GET"
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  })
), fetchList = (client, type, page, perPage) => defer(
  () => client.observable.request({
    url: `/addons/unsplash/photos?order_by=${type}&page=${page}&per_page=${perPage}`,
    withCredentials: !0,
    method: "GET"
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  })
);
function fetchDownloadUrl(client, photo) {
  const downloadUrl = photo.links.download_location.replace(
    "https://api.unsplash.com",
    "/addons/unsplash"
  );
  return client.request({
    url: downloadUrl,
    withCredentials: !0,
    method: "GET"
  }).then((result) => result.url);
}
const search = (client, query, page, resultsPerPage) => concat(
  query.pipe(
    withLatestFrom(page),
    debounceTime(500),
    distinctUntilChanged(),
    switchMap(([q, p]) => q ? fetchSearch(client, q, p, resultsPerPage).pipe(
      distinctUntilChanged(),
      map((result) => result.results)
    ) : fetchList(client, "popular", p, resultsPerPage))
  )
), Root = styled.div`
  overflow: hidden;
  background-origin: content-box;
  background-repeat: no-repeat;
  background-clip: border-box;
  background-size: cover;
  position: relative;
  outline: none !important;
  border: ${({ theme }) => `1px solid ${theme.sanity.color.card.enabled.border}`};
  box-sizing: content-box;
  user-drag: none;

  &:hover {
    opacity: 0.85;
  }

  &:focus,
  &:active {
    border: 1px solid var(--input-border-color-focus);
    box-shadow: inset 0 0 0 3px var(--input-border-color-focus);
  }
`, CreditLineLink = styled.a`
  text-decoration: none;
  cursor: pointer;
`, CreditLine = styled(Card)`
  ${({ theme }) => `
     --creditline-fg: ${theme.sanity.color.card.enabled.fg};
     --creditline-bg: ${theme.sanity.color.card.enabled.bg};
   `};
  user-drag: none;
  position: absolute;
  background-color: var(--creditline-bg);
  bottom: 0;

  [data-ui='Text'] {
    color: var(--creditline-fg);
  }
`, UTM_SOURCE = "sanity-plugin-asset-source-unsplash";
function Photo(props) {
  const { onClick, data, onKeyDown, onFocus, active, width, height } = props, handleClick = useCallback(() => {
    onClick(data);
  }, [onClick, data]), handleCreditLineClicked = useCallback(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (event) => {
      event.stopPropagation();
      const url = `${data.links.html}?utm_source=${encodeURIComponent(
        UTM_SOURCE
      )}&utm_medium=referral`;
      window.open(url, data.id, "noreferrer,noopener");
    },
    [data]
  ), handleKeyDown = useCallback(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (event) => {
      onKeyDown(event), event.keyCode === 13 && onClick(data);
    },
    [onKeyDown, data, onClick]
  ), handleMouseDown = useCallback(() => {
    onFocus(data);
  }, [onFocus, data]);
  useEffect(() => {
    active && onFocus(data);
  }, [active, data, onFocus]);
  const src = data.urls.small, userName = data.user.name;
  return /* @__PURE__ */ jsx(
    Root,
    {
      title: `Select image by ${userName} from Unsplash`,
      tabIndex: 0,
      onKeyDown: handleKeyDown,
      onMouseDown: handleMouseDown,
      style: {
        width: `${width}px`,
        height: `${height}px`,
        backgroundImage: `url("${src}")`
      },
      onClick: handleClick,
      children: /* @__PURE__ */ jsx(CreditLineLink, { onClick: handleCreditLineClicked, children: /* @__PURE__ */ jsx(CreditLine, { padding: 2, radius: 2, margin: 2, children: /* @__PURE__ */ jsxs(Text, { size: 1, title: `Open image by ${userName} on Unsplash in new window`, children: [
        "By @",
        data.user.username
      ] }) }) })
    }
  );
}
const SearchInput = styled(TextInput)`
  position: sticky;
  top: 0;
  z-index: 1;
`;
styled.div`
  overflow-y: auto;
  max-height: 80vh;
`;
var __defProp = Object.defineProperty, __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: !0, configurable: !0, writable: !0, value }) : obj[key] = value, __publicField = (obj, key, value) => (__defNormalProp(obj, typeof key != "symbol" ? key + "" : key, value), value);
const RESULTS_PER_PAGE = 42, PHOTO_SPACING = 2, PHOTO_PADDING = 1;
function UnsplashAssetSource(props) {
  const client = useClient({ apiVersion: "2022-09-01" });
  return /* @__PURE__ */ jsx(UnsplashAssetSourceInternal, { ...props, client });
}
class UnsplashAssetSourceInternal extends React.Component {
  constructor() {
    super(...arguments), __publicField(this, "state", {
      cursor: 0,
      query: "",
      page: 1,
      searchResults: [[]],
      isLoading: !0
    }), __publicField(this, "searchSubscription", null), __publicField(this, "searchSubject$", new BehaviorSubject("")), __publicField(this, "pageSubject$", new BehaviorSubject(1)), __publicField(this, "handleSelect", (photo) => (this.setState({ isLoading: !0 }), fetchDownloadUrl(this.props.client, photo).then((downloadUrl) => {
      const description = photo.description || void 0, asset = {
        kind: "url",
        value: downloadUrl,
        assetDocumentProps: {
          _type: "sanity.imageAsset",
          source: {
            name: "unsplash",
            id: photo.id,
            url: photo.links.html
          },
          description,
          creditLine: `${photo.user.name} by Unsplash`
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
        }
      };
      this.props.onSelect([asset]);
    }))), __publicField(this, "handleClose", () => {
      this.props.onClose();
    }), __publicField(this, "handleSearchTermChanged", (event) => {
      const query = event.currentTarget.value;
      this.setState({ query, page: 1, searchResults: [[]], isLoading: !0, cursor: 0 }), this.pageSubject$.next(1), this.searchSubject$.next(query);
    }), __publicField(this, "handleSearchTermCleared", () => {
      this.setState({ query: "", page: 1, searchResults: [[]], isLoading: !0, cursor: 0 }), this.pageSubject$.next(1), this.searchSubject$.next("");
    }), __publicField(this, "handleScollerLoadMore", () => {
      const nextPage = this.state.page + 1;
      this.setState({ page: nextPage, isLoading: !0 }), this.pageSubject$.next(nextPage), this.searchSubject$.next(this.state.query);
    }), __publicField(this, "handleKeyDown", (event) => {
      const { cursor } = this.state;
      (event.keyCode === 38 || event.keyCode === 37) && cursor > 0 ? this.setState((prevState) => ({
        cursor: prevState.cursor - 1
      })) : (event.keyCode === 40 || event.keyCode === 39) && cursor < this.getPhotos().length - 1 && this.setState((prevState) => ({
        cursor: prevState.cursor + 1
      }));
    }), __publicField(this, "updateCursor", (photo) => {
      const index = this.getPhotos().findIndex((result) => result.id === photo.id);
      this.setState({ cursor: index });
    }), __publicField(this, "renderImage", (props) => {
      const { photo, layout } = props, active = this.getPhotos().findIndex((result) => result.id === photo.data.id) === this.state.cursor || !1;
      return /* @__PURE__ */ jsx(
        Photo,
        {
          onClick: this.handleSelect.bind(photo.data),
          onKeyDown: this.handleKeyDown,
          data: photo.data,
          width: layout.width,
          height: layout.height,
          active,
          onFocus: this.updateCursor
        }
      );
    });
  }
  componentDidMount() {
    this.searchSubscription = search(
      this.props.client,
      this.searchSubject$,
      this.pageSubject$,
      RESULTS_PER_PAGE
    ).subscribe({
      next: (results) => {
        this.setState((prev) => ({
          searchResults: [...prev.searchResults, results],
          isLoading: !1
        }));
      }
    });
  }
  componentWillUnmount() {
    this.searchSubscription && this.searchSubscription.unsubscribe();
  }
  getPhotos() {
    return this.state.searchResults.flat();
  }
  render() {
    const { query, searchResults, isLoading } = this.state;
    return /* @__PURE__ */ jsx(
      Dialog,
      {
        animate: !0,
        id: "unsplash-asset-source",
        header: "Select image from Unsplash",
        onClose: this.handleClose,
        open: !0,
        width: 4,
        children: /* @__PURE__ */ jsxs(Stack, { space: 3, paddingX: 4, paddingBottom: 4, children: [
          /* @__PURE__ */ jsx(
            SearchInput,
            {
              clearButton: query.length > 0,
              icon: SearchIcon,
              onChange: this.handleSearchTermChanged,
              onClear: this.handleSearchTermCleared,
              placeholder: "Search by topics or colors",
              value: query
            }
          ),
          !isLoading && this.getPhotos().length === 0 && /* @__PURE__ */ jsx(Text, { size: 1, muted: !0, children: "No results found" }),
          /* @__PURE__ */ jsx(
            InfiniteScroll,
            {
              dataLength: this.getPhotos().length,
              next: this.handleScollerLoadMore,
              hasMore: !0,
              scrollThreshold: 0.99,
              height: "60vh",
              loader: /* @__PURE__ */ jsx(Flex, { align: "center", justify: "center", padding: 3, children: /* @__PURE__ */ jsx(Spinner, { muted: !0 }) }),
              endMessage: /* @__PURE__ */ jsx(Text, { size: 1, muted: !0, children: "No more results" }),
              children: searchResults.filter((photos) => photos.length > 0).map((photos, index) => /* @__PURE__ */ jsx(
                PhotoAlbum,
                {
                  layout: "rows",
                  spacing: PHOTO_SPACING,
                  padding: PHOTO_PADDING,
                  targetRowHeight: (width) => width < 300 ? 150 : width < 600 ? 200 : 300,
                  photos: photos.map((photo) => ({
                    src: photo.urls.small,
                    width: photo.width,
                    height: photo.height,
                    key: photo.id,
                    data: photo
                  })),
                  renderPhoto: this.renderImage,
                  componentsProps: {
                    containerProps: { style: { marginBottom: `${PHOTO_SPACING}px` } }
                  }
                },
                `gallery-${query || "popular"}-${index}`
              ))
            }
          )
        ] })
      }
    );
  }
}
__publicField(UnsplashAssetSourceInternal, "defaultProps", {
  selectedAssets: void 0
});
const unsplashAssetSource = {
  name: "unsplash",
  title: "Unsplash",
  component: UnsplashAssetSource,
  icon: UnsplashIcon
}, unsplashImageAsset = definePlugin({
  name: "asset-source-unsplash-plugin",
  form: {
    image: {
      assetSources: (prev) => [...prev, unsplashAssetSource]
    }
  }
});
export {
  unsplashAssetSource,
  unsplashImageAsset
};
